### PR TITLE
[SYCL] Change which constructor is gated for devices

### PR DESCRIPTION
**Pull request by jbrodman**
SYCL 1.2.1 includes a constructor for `multi_ptr` from `T*`, available on both the host and device.  The current implementation disables this constructor on the host to avoid conflicting definitions for constructors using `pointer_t` and `T*`.

Making this constructor available enables implicit conversions between pointer types and `multi_ptr` classes, and addresses requests such as #849. 

Signed-off-by: James Brodman <james.brodman@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
@jbrodman, please, ignore failure in clang-format-check job. It seems to include irrelevant changes into the validated change.

**Comment by jbrodman** [comment]
> @jbrodman, please, ignore failure in clang-format-check job. It seems to include irrelevant changes into the validated change.

👍

File: sycl/include/CL/sycl/multi_ptr.hpp [hunk 1]
```diff
@@ -44,13 +44,15 @@ template <typename ElementType, access::address_space Space> class multi_ptr {
   multi_ptr() : m_Pointer(nullptr) {}
   multi_ptr(const multi_ptr &rhs) = default;
   multi_ptr(multi_ptr &&) = default;
-  multi_ptr(pointer_t pointer) : m_Pointer(pointer) {}
 #ifdef __SYCL_DEVICE_ONLY__
+  multi_ptr(pointer_t pointer) : m_Pointer(pointer) {}
+#endif
```
Comment (bader) [inline sycl/include/CL/sycl/multi_ptr.hpp]: I think we should apply the same change to `operator=`.

Comment (jbrodman) [reply sycl/include/CL/sycl/multi_ptr.hpp]: Done.

**Pull request merged**